### PR TITLE
Additional fix on 21801

### LINKED ISSUE #21801: Moving Folders
Folders should always have the knowable/consistant id.

If someone wants to "move" a folder, we should not update the folder record. Instead, we should create a new folder record that has the correct id generated and delete the old folder / identifier when done.

I believe we will also need to update the info in the permission/permission_reference table as well.

### PR DESCRIPTION
**Pull request by nollymar**
On this PR: 
* FolderAPI.exists(String) can filter by folderId or folderInode. A test was implemented.
* Fixing JS exception in SiteBrowser when you double click on a folder created on a previous dotCMS version.
* Fixing ClassCastException when a folder containing a FileAsset is moved.
* Copying/Renaming a folder must generate a new identifier. Tests were updated to support the new logic.
* Serialization exception when a MenuLink is added to a folder.
* Method `validateFolderName` was moved from `FolderFactoryImpl` to `FolderAPI`.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/portlets/folders/action/EditFolderAction.java [hunk 1]
```diff
@@ -360,33 +365,33 @@ public boolean _updateFolder(ActionRequest req, ActionResponse res,PortletConfig
 			}
 
 			if (parentFolder instanceof Folder){
-				if(f.getName().equalsIgnoreCase("admin")){
+				if(folder.getName().equalsIgnoreCase("admin")){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/folders/action/EditFolderAction.java:368]: use the constant in the left

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java [hunk 2]
```diff
@@ -761,39 +761,39 @@ else if(cont.getStructure().getStructureType()==Structure.STRUCTURE_TYPE_HTMLPAG
     @SuppressWarnings("unchecked")
 	private List<Map> getFoldersTree (Folder parent, Role[] roles) throws DotStateException, DotDataException, DotSecurityException {
         FolderAPI folderAPI = APILocator.getFolderAPI();
-        List<Folder> children = new ArrayList<Folder>();
+        List<Folder> children = new ArrayList<>();
 		try {
 			children = folderAPI.findSubFolders(parent,userAPI.getSystemUser(),false);
 		} catch (Exception e) {
 			Logger.error(this, "Could not load folders : ",e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java:768]: Let's log the ID of the parent folder whose tree could not be loaded

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java [hunk 3]
```diff
@@ -761,39 +761,39 @@ else if(cont.getStructure().getStructureType()==Structure.STRUCTURE_TYPE_HTMLPAG
...
         		folderMap.put("selected", false);
-        	folderMap.put("parent", parentInode);
+        	folderMap.put("parent", parentId);
 
         	List permissions = new ArrayList();
         	try {
-        		permissions = permissionAPI.getPermissionIdsFromRoles(f, roles, usr);
+        		permissions = permissionAPI.getPermissionIdsFromRoles(folder, roles, usr);
         	} catch (DotDataException e) {
         		Logger.error(this, "Could not load permissions : ",e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java:798]: Let's log the ID of the parent folder here too.

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java [hunk 4]
```diff
@@ -1941,24 +1941,24 @@ private long getContentSelectedLanguageId(HttpServletRequest request){
...
 				List<Folder> subFolders = folderAPI.findSubFolders(host,user,false);
-				for (Folder f : subFolders) {
+				for (final Folder folder : subFolders) {
 
 						List permissions = new ArrayList();
 						try {
-							permissions = permissionAPI.getPermissionIdsFromRoles(f, roles, user);
+							permissions = permissionAPI.getPermissionIdsFromRoles(folder, roles, user);
 						} catch (DotDataException e) {
 							Logger.error(this, "Could not load permissions : ",e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java:1950]: Let's log the ID of the folder here too.

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]

**Review by fabrizzio-dotCMS — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java (hunk 4 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java (hunk 3 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java (hunk 2 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/folders/action/EditFolderAction.java (hunk 1 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/portlets/folders/action/EditFolderAction.java]: Done

**Review by jcastro-dotcms — APPROVED** [review]

**Pull request merged**